pkg/operator/ceph/cluster/watcher.go

```diff
@@ -130,6 +170,122 @@ func (c *clientCluster) onK8sNode(ctx context.Context, object runtime.Object) bo
 	return false
 }
 
+func (c *clientCluster) createNetworkFence(ctx context.Context, node corev1.Node, clusterName string) error {
+	nodeIPtoBlock, pv, err := c.getWatcherIpPortToBlock(ctx, node, clusterName)
+	if err != nil {
+		return fmt.Errorf("failed to get node IP to block for node %q. %v", node.Name, err)
+	}
+	fmt.Println("NodeIP to block", nodeIPtoBlock)
+
+	networkFence := &addonsv1alpha1.NetworkFence{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      node.Name,
+			Namespace: c.namespace,
+		},
+		Spec: addonsv1alpha1.NetworkFenceSpec{
+			Driver:     "rook-ceph.rbd.csi.ceph.com",
+			FenceState: addonsv1alpha1.Fenced,
+			Secret: addonsv1alpha1.SecretSpec{
+				Name:      pv.Spec.CSI.ControllerPublishSecretRef.Name,
+				Namespace: pv.Spec.CSI.ControllerPublishSecretRef.Namespace,
+			},
+			Cidrs: []string{
+				nodeIPtoBlock,
+			},
+			Parameters: map[string]string{
+				"clusterID": pv.Spec.CSI.VolumeAttributes["clusterID"],
+			},
+		},
+	}
+
+	err = c.client.Create(ctx, networkFence)
+	if err != nil {
+		if errors.IsAlreadyExists(err) {
+			logger.Infof("network fence CR %s already exists", node.Name)
+		} else {
+			logger.Errorf("failed to create network fence CR. %v", err)
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *clientCluster) deleteNetworkFence(ctx context.Context, node corev1.Node) error {
+	err := c.client.Delete(ctx, &addonsv1alpha1.NetworkFence{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      node.Name,
+			Namespace: c.namespace,
+		},
+	})
+	if err != nil && !errors.IsNotFound(err) {
+		logger.Errorf("failed to delete network fence CR. %v", err)
+		return err
+	}
+	return nil
+}
+
+func (c *clientCluster) getWatcherIpPortToBlock(ctx context.Context, node corev1.Node, clusterName string) (string, corev1.PersistentVolume, error) {
+	volumesInuse := node.Status.VolumesInUse
+	logger.Debugf("volumesInuse %s", volumesInuse)
+
+	listPVs, err := c.context.Clientset.CoreV1().PersistentVolumes().List(context.TODO(), metav1.ListOptions{})
```

Comment:
We are listing all PVs in the entire cluster? Is there no way to limit the query to the node? There could be thousands of PVs in the cluster.